firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalDocumentsView.java

```diff
@@ -317,36 +327,208 @@ void recalculateAndSaveOverlays(Set<DocumentKey> documentKeys) {
     }
   }
 
-  /** Queries the remote documents and overlays by doing a full collection scan. */
+  /**
+   * Given a collection group, returns the next documents that follow the provided offset, along
+   * with an updated offset.
+   *
+   * <p>The first call to this method should set {@code updateMemoizedResults} to {@code true} to
+   * load remote documents and mutations into persistence. Subsequent calls to this method should be
+   * made with {@code updateMemoizedResults} set to {@code false} {@code offset} set to the offset
+   * returned by the previous call.
+   *
+   * <p>This method returns documents in order of read time from the provided offset, followed by
+   * documents with mutations in order of batch id from the offset.
+   *
+   * <p>If no documents are found, returns an empty list and an offset with the latest read time in
+   * the remote document cache.
+   *
+   * @param collectionGroup The collection group for the documents.
+   * @param offset The offset to index into.
+   * @param updateMemoizedResults Whether to load documents and mutations from persistence.
+   * @return A pair containing the next offset and a list of documents that follow the provided
+   *     offset.
+   */
+  public Pair<IndexOffset, List<Document>> getNextDocumentsAndOffsetForCollectionGroup(
+      String collectionGroup, IndexOffset offset, boolean updateMemoizedResults) {
+
+    if (updateMemoizedResults) {
```

Comment:
I am not 100% convinced that we need to memoize results (and if we do need to do so, we need to discuss what this means if the user changes).

My expected code flow would be as such:

- Fetch 50 documents from RDC.
- If less than 50 documents, fetch next documents from MutationQueue.
- Combine the two results by applying mutations
- If we really care, we can fetch more documents if we are still below 50.

Reply:
Implemented.